design/vela-core/rollout-design.md

```diff
@@ -0,0 +1,269 @@
+	// +optional
+	SourceRef []runtimev1alpha1.TypedReference `json:"sourceRef,omitempty"`
+
+	// RolloutPlan is the details on how to rollout the resources
+	RolloutPlan RolloutPlan `json:"rolloutPlan"`
+}
+```
+The target and source here are the same as other OAM traits that refers to a workload instance
+ with its GVK and name. 
+ 
+We can see that the core part of the rollout logic is encapsulated by the `RolloutPlan` object. This
+ allows us to write multiple controllers that share the same logic without duplicating the code
+ . Here is the definition of the `RolloutPlan` structure.
+
+```go
+// RolloutPlan fines the details of the rollout plan
+type RolloutPlan struct {
+
+	// RolloutStrategy defines strategies for the rollout plan
+	// +optional
+	RolloutStrategy RolloutStrategyType `json:"rolloutStrategy,omitempty"`
+
+	// The size of the target resource. The default is the same
+	// as the size of the source resource.
+	// +optional
+	TargetSize *int32 `json:"targetSize,omitempty"`
+
+	// The number of batches, default = 1
+	// mutually exclusive to RolloutBatches
+	// +optional
+	NumBatches *int32 `json:"numBatches,omitempty"`
+
+	// The exact distribution among batches.
+	// mutually exclusive to NumBatches
+	// +optional
+	RolloutBatches []RolloutBatch `json:"rolloutBatches,omitempty"`
+
+	// All pods in the batches up to the batchPartition (included) will have
+	// the target resource specification while the rest still have the source resource
+	// This is designed for the operators to manually rollout
+	// Default is the the number of batches which will rollout all the batches
+	// +optional
+	BatchPartition *int32 `json:"lastBatchToRollout,omitempty"`
+
+	// RevertOnDelete revert the rollout when the rollout CR is deleted, default is false 
+	//+optional 
+	RevertOnDelete bool `json:"revertOnDelete,omitempty"`
+
+	// Paused the rollout, default is false 
+	//+optional 
+	Paused bool `json:"paused,omitempty"`
+
+	// RolloutWebhooks provides a way for the rollout to interact with an external process
+	// +optional
+	RolloutWebhooks []RolloutWebhook `json:"rolloutWebhooks,omitempty"`
+
+	// CanaryMetric provides a way for the rollout process to automatically check certain metrics
+	// before complete the process
+	// +optional
+	CanaryMetric []CanaryMetric `json:"canaryMetric,omitempty"`
+}
+```
+
+## User Experience Workflow
+OAM rollout experience is different from flagger in a few different ways and here are the
+ implications on its impact on the user experience.
+- We assume that the resources it refers to are **_immutable_**. In contrast, flagger watches over a
+  target resource and reacts whenever the target's specification changes.
+    - The trait version of the controller refers to componentRevision.
+    - The application version of the controller refers to immutable application. 
+- The rollout logic **_works only once_** and stops after it reaches a terminal state. One can
```

Comment:
Does it mean if a rollout CR works completely, then it won't work even the CR spec is changed?

Reply:
the only way to trigger it to restart is when the target or source resource is changed